p2p/background/router_test.go

```diff
@@ -0,0 +1,355 @@
+	testRouter := newTestRouter(t, nil)
+	peerstore := testRouter.host.Peerstore()
+
+	// NB: assert initial state
+	require.Len(t, testRouter.pstore, 1)
+
+	existingPeer := testRouter.pstore.GetPeerList()[0]
+	require.NotNil(t, existingPeer)
+
+	existingPeerInfo, err := utils.Libp2pAddrInfoFromPeer(existingPeer)
+	require.NoError(t, err)
+
+	existingPeerstoreAddrs := peerstore.Addrs(existingPeerInfo.ID)
+	require.Len(t, existingPeerstoreAddrs, 1)
+
+	existingPeerMultiaddr, err := utils.Libp2pMultiaddrFromServiceURL(existingPeer.GetServiceURL())
+	require.NoError(t, err)
+	require.Equal(t, existingPeerstoreAddrs[0].String(), existingPeerMultiaddr.String())
+
+	err = testRouter.RemovePeer(existingPeer)
+	require.NoError(t, err)
+
+	require.Len(t, testRouter.pstore, 0)
+
+	// NB: libp2p peerstore implementations only remove peer keys and metadata
+	// but continue to resolve multiaddrs until their respective TTLs expire.
+	// (see: https://github.com/libp2p/go-libp2p/blob/v0.25.1/p2p/host/peerstore/pstoremem/peerstore.go#L108)
+	// (see: https://github.com/libp2p/go-libp2p/blob/v0.25.1/p2p/host/peerstore/pstoreds/peerstore.go#L187)
+
+	existingPeerstoreAddrs = peerstore.Addrs(existingPeerInfo.ID)
+	require.Len(t, existingPeerstoreAddrs, 1)
+}
+
+func TestBackgroundRouter_Broadcast(t *testing.T) {
+	const (
+		numPeers            = 4
+		testMsg             = "test messsage"
+		testTimeoutDuration = time.Second * 5
+	)
+
+	var (
+		ctx = context.Background()
+		mu  sync.Mutex
+		// map used as a set to collect IDs of peers which have received a message
+		seenMessages = make(map[string]struct{})
+		wg           = sync.WaitGroup{}
+		done         = make(chan struct{}, 1)
+		testTimeout  = time.After(testTimeoutDuration)
+		// NB: peerIDs are stringified
+		actualPeerIDs   []string
+		expectedPeerIDs = make([]string, numPeers)
+		testHosts       = make([]libp2pHost.Host, 0)
+		libp2pMockNet   = mocknet.New()
+	)
+
+	// setup 4 libp2p hosts to listen for incoming streams from the test backgroundRouter
+	for i := 0; i < numPeers; i++ {
+		wg.Add(1)
+
+		privKey, selfPeer := newTestPeer(t)
+		host := newTestHost(t, libp2pMockNet, privKey)
+		testHosts = append(testHosts, host)
+		expectedPeerIDs[i] = host.ID().String()
+		rtr := newRouterWithSelfPeerAndHost(t, selfPeer, host)
+		go readSubscription(t, ctx, &wg, rtr, &mu, seenMessages)
+	}
+
+	// bootstrap off of arbitrary testHost
+	privKey, selfPeer := newTestPeer(t)
+
+	// set up a test backgroundRouter
+	testRouterHost := newTestHost(t, libp2pMockNet, privKey)
+	testRouter := newRouterWithSelfPeerAndHost(t, selfPeer, testRouterHost)
+	testHosts = append(testHosts, testRouterHost)
+
+	// simulate network links between each to every other
+	// (i.e. fully-connected network)
+	err := libp2pMockNet.LinkAll()
+	require.NoError(t, err)
+
+	bootstrap(t, ctx, testHosts)
+
+	go func() {
+		// wait for hosts to listen and peer discovery
+		time.Sleep(time.Second * 2)
```

Comment:
If possible, I would prefer waiting on a signal channel instead to prevent a load bearing sleep.

Reply:
I did my best; with the addition of notify/notifee I was able to reduce it to ~100ms on my machine, so I set it to 250ms to be on the safe side.